ngsPETSc/utils/firedrake/adaptive_hierarchy.py

```diff
@@ -0,0 +1,331 @@
+    """
+    _ = FunctionSpace(mesh, "DG", 0)
+    _ = FunctionSpace(coarse_mesh, "DG", 0)
+    ngmesh = mesh.netgen_mesh
+    num_parents = coarse_mesh.num_cells()
+
+    if mesh.topology_dm.getDimension() == 2:
+        parents = ngmesh.parentsurfaceelements.NumPy()
+        elements = ngmesh.Elements2D()
+    else:
+        parents = ngmesh.parentelements.NumPy()
+        elements = ngmesh.Elements3D()
+
+    c2f = [[] for _ in range(num_parents)]
+    f2c = [[] for _ in range(mesh.num_cells())]
+
+    if parents.shape[0] == 0:
+        raise RuntimeError("Added mesh has not refined any cells from previous mesh")
+    for l, _ in enumerate(elements):
+        if parents[l][0] == -1 or l < num_parents:
+            f2c[mesh._cell_numbering.getOffset(l)].append(coarse_mesh._cell_numbering.getOffset(l))
+            c2f[coarse_mesh._cell_numbering.getOffset(l)].append(mesh._cell_numbering.getOffset(l))
+
+        elif parents[l][0] < num_parents:
+            fine_ind = mesh._cell_numbering.getOffset(l)
+            coarse_ind = coarse_mesh._cell_numbering.getOffset(parents[l][0])
+            f2c[fine_ind].append(coarse_ind)
+            c2f[coarse_ind].append(fine_ind)
+
+        else:
+            a = parents[parents[l][0]][0]
+            while a >= num_parents:
+                a = parents[a][0]
+
+            f2c[mesh._cell_numbering.getOffset(l)].append(coarse_mesh._cell_numbering.getOffset(a))
+            c2f[coarse_mesh._cell_numbering.getOffset(a)].append(mesh._cell_numbering.getOffset(l))
+
+    return c2f, np.array(f2c).astype(int)
+
+
+def split_to_submesh(mesh, coarse_mesh, c2f, f2c):
+    """
+    Computes submesh split from full mesh
+    """
+    if mesh.topological_dimension() <= 2:
+        max_children = 4
+    else:
+        max_children = 16
+    V = FunctionSpace(mesh, "DG", 0)
+    V2 = FunctionSpace(coarse_mesh, "DG", 0)
+    coarse_splits = {
+        i: Function(V2, name=f"{i}_elements") for i in range(1, max_children + 1)
+    }
+    fine_splits = {
+        i: Function(V, name=f"{i}_elements") for i in range(1, max_children + 1)
+    }
+    num_children = np.zeros((len(c2f)))
+
+    for i, children in enumerate(c2f):
+        n = len(children)
+        if 1 <= n <= max_children:
+            coarse_splits[n].dat.data[i] = 1
+            num_children[i] = n
+
+    for i in range(1, max_children + 1):
+        fine_splits[i].dat.data[num_children[f2c.squeeze()] == i] = 1
+
+    return coarse_splits, fine_splits, num_children
+
+
+def full_to_sub(mesh, submesh, label):
+    """
+    Returns the submesh element id associated with the full mesh element id
+    """
+    V1 = FunctionSpace(mesh, "DG", 0)
+    V2 = FunctionSpace(submesh, "DG", 0)
+    u1 = Function(V1)
+    u2 = Function(V2)
+    u2.dat.data[:] = np.arange(len(u2.dat.data))
+    u1.interpolate(u2, subset=mesh.cell_subset(label))
+
+    return lambda x: u1.dat.data[x].astype(int)
```

Comment:
It'd be better to return the numpy array and index it on a list as opposed to calling this lambda multiple times